modules/core/04-channel/types/upgrade.go

```diff
@@ -77,12 +71,36 @@ func (ut Timeout) IsValid() bool {
 	return !ut.Height.IsZero() || ut.Timestamp != 0
 }
 
-// NewErrorReceipt returns an error receipt with the code from the provided error type stripped
-// out to ensure changes of the error message don't cause state machine breaking changes.
-func NewErrorReceipt(upgradeSequence uint64, err error) ErrorReceipt {
-	_, code, _ := errorsmod.ABCIInfo(err, false) // discard non-determinstic codespace and log values
+// UpgradeError defines an error that occurs during an upgrade.
+type UpgradeError struct {
+	// err is the underlying error that caused the upgrade to fail.
+	// this error should not be written to state.
+	err error
+	// sequence is the upgrade sequence number of the upgrade that failed.
+	sequence uint64
+}
+
+func (u UpgradeError) Error() string {
+	return u.err.Error()
+}
+
+// GetErrorReceipt returns an error receipt with the code from the underlying error type stripped.
+func (u UpgradeError) GetErrorReceipt() ErrorReceipt {
+	// restoreErrorString defines a string constant included in error receipts.
+	// NOTE: Changing this const is state machine breaking as it is written into state.
+	const restoreErrorString = "restored channel to pre-upgrade state"
```

Comment:
moved this into the function to make it more obvious that this should not be changed if this function is modified in the future.

Reply:
agree, makes more sense here imo!